### PR TITLE
feat(attachSession): attach protractor to existing webdriver session

### PR DESCRIPTION
**Pull request by cnishina**
Attaching an existing selenium browser session to protractor rather than always creating new one. The session can be passed into the config file as a string via the `sessionId`.

### CODE REVIEW

File: lib/configParser.js [hunk 1]
```diff
@@ -33,6 +33,7 @@ var ConfigParser = function() {
       defaultTimeoutInterval: (30 * 1000)
     },
     seleniumArgs: [],
+    sessionId: null,
```
Comment (juliemr) [inline lib/configParser.js]: Can you also update our master list of configuration docs at https://github.com/angular/protractor/blob/master/docs/referenceConf.js?

Comment (juliemr) [inline lib/configParser.js]: Let's rename to `seleniumSessionId` - to err on the side of being too verbose but obvious what this belongs to.

File: lib/driverProviders/attachSession.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ */
+AttachedSessionDriverProvider.prototype.setupEnv = function() {
+  log.puts('Using the selenium server at ' + this.config_.seleniumAddress);
+  log.puts('Using session id - ' + this.config_.sessionId);
+  return q.fcall(function() {});
+};
+
+
+/**
+ * Getting a new driver by attaching an existing session
```
Comment (juliemr) [inline lib/driverProviders/attachSession.js]: Nit: always end comments with a '.'

File: lib/driverProviders/attachSession.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * @public
+ * @return webdriver instance
+ */
+AttachedSessionDriverProvider.prototype.getNewDriver = function() {
+  var executor = executors.createExecutor(this.config_.seleniumAddress, null);
```
Comment (juliemr) [inline lib/driverProviders/attachSession.js]: Nit: Looks like we can just do `executors.createExecutor(this.config_.seleniumAddress)`, no need for the last null parameter.

File: lib/driverProviders/attachSession.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  var newDriver = new webdriver.WebDriver
+      .attachToSession(executor, this.config_.sessionId);
```
Comment (juliemr) [inline lib/driverProviders/attachSession.js]: What's the error message look like if the session id is invalid? Is it reasonably actionable for a user, or do we need to wrap it?

File: scripts/attachSessionHelper.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  }
+};
+var TEST = {};
+TEST.body = '';
+
+var setBody = function(chunk) {
+  TEST.body += chunk.toString();
+  var sessionId = JSON.parse(TEST.body).sessionId;
+  console.log('created new session with id: ' + sessionId);
+  fs.writeFile('spec/attachSessionTemp.js',
```
Comment (juliemr) [inline scripts/attachSessionHelper.js]: I think it would be cleaner to use Node's [child_process](https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback) to start up protractor and add a command line flag, e.g. `bin/protractor spec/attachSessionConf.js --sessionId=***`

Then we don't have to mess with the file system.

Comment (juliemr) [inline scripts/attachSessionHelper.js]: And, if we use child process, we can intelligently wait for protractor to be done instead of using an arbitrary timeout.

File: lib/driverProviders/attachSession.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  this.drivers_.push(newDriver);
+  return newDriver;
+
+};
+
+
```
Comment (juliemr) [inline lib/driverProviders/attachSession.js]: I think we'll need to override `quitDriver` as well to avoid killing the session when we're done with it.

File: scripts/test.js [hunk 7]
```diff
@@ -26,16 +26,17 @@ var passingTests = [
   'node lib/cli.js spec/plugins/browserGetUnsyncedConf.js',
   'node lib/cli.js spec/plugins/waitForAngularConf.js',
   'node lib/cli.js spec/interactionConf.js',
-  'node lib/cli.js spec/directConnectConf.js',
+  //'node lib/cli.js spec/directConnectConf.js',
```
Comment (juliemr) [inline scripts/test.js]: These are working on my machine - lets leave them in, and we can figure out why your machine doesn't like them (or maybe we're missing some setup instructions that I've forgotten about).

File: scripts/attachSessionHelper.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    'var env = require(\'./environment.js\'); \n' +
+    '\n' +
+    'exports.config = { \n' +
+    '  seleniumAddress: env.seleniumAddress, \n' +
+    '  sessionId: \'' + sessionId + '\', \n' +
+    '  framework: \'jasmine\', \n' +
+    '  specs: [\'attachSessionProvider/*_spec.js\'], \n' +
+    '  baseUrl: \'http://localhost:8081\' \n' +
+    '};\n');
+  setTimeout(function(){}, 3000);
```
Comment (juliemr) [inline scripts/attachSessionHelper.js]: We should send a command to quit the session after protractor finishes, as well.

File: spec/attachSessionProvider/attachSession_spec.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+describe('attach session connect', function() {
+  var URL = '/ng2/#/async';
+
+  beforeEach(function() {
+    browser.get(URL);
+  });
+  it('create a new session', function() {
```
Comment (juliemr) [inline spec/attachSessionProvider/attachSession_spec.js]: Nit: `creates` - specs read like a sentence; "attach session connect creates a new session" :)

**Comment by juliemr** [comment]
Added some review comments!

File: scripts/test.js (hunk 7 above)
Comment (cnishina) [inline scripts/test.js]: This didn't run for me because it is using chromedriver directly. My error is: Could not find chromedriver at protractor/selenium/chromedriver

File: lib/driverProviders/attachSession.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+var AttachedSessionDriverProvider = function(config) {
+  DriverProvider.call(this, config);
+};
+util.inherits(AttachedSessionDriverProvider, DriverProvider);
+
+/**
+ * Configure and launch (if applicable) the object's environment.
+ * @public
+ * @return {q.promise} A promise which will resolve when the environment is
```
Comment (sjelin) [inline lib/driverProviders/attachSession.js]: nit: q.promise<function>

File: lib/driverProviders/attachSession.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+ *     ready to test.
+ */
+AttachedSessionDriverProvider.prototype.setupEnv = function() {
+  log.puts('Using the selenium server at ' + this.config_.seleniumAddress);
+  log.puts('Using session id - ' + this.config_.sessionId);
+  return q.fcall(function() {});
```
Comment (sjelin) [inline lib/driverProviders/attachSession.js]: Why not `q(undefined)`?

File: lib/driverProviders/attachSession.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+};
+
+
+/**
+ * Getting a new driver by attaching an existing session
+ *
+ * @public
+ * @return webdriver instance
```
Comment (sjelin) [inline lib/driverProviders/attachSession.js]: Needs type `{webdriver.WebDriver}`

File: scripts/attachSession.sh [hunk 13]
```diff
@@ -0,0 +1,2 @@
+node scripts/attachSessionHelper.js
```
Comment (sjelin) [inline scripts/attachSession.sh]: I think we should put this & attachSessionHelper in a subfolder

File: lib/driverProviders/attachSession.js (hunk 4 above)
Comment (cnishina) [inline lib/driverProviders/attachSession.js]: The error message when the session id is invalid returns an `UnkonwnError`. The error is being caused in the selenium/webdriver.js and I cannot catch the error.

File: lib/driverProviders/attachSession.js [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Configure and launch (if applicable) the object's environment.
+ * @public
+ * @return {q.promise} A promise which will resolve when the environment is
+ *     ready to test.
+ */
+AttachedSessionDriverProvider.prototype.setupEnv = function() {
+  log.puts('Using the selenium server at ' + this.config_.seleniumAddress);
+  log.puts('Using session id - ' + this.config_.seleniumSessionId);
+  return q.fcall(function() {});
```
Comment (sjelin) [inline lib/driverProviders/attachSession.js]: Why not `q(undefined)`?

Comment (cnishina) [inline lib/driverProviders/attachSession.js]: I could change that to `q(undefined)`. I was basing the `setupEnv` off of hosted.js.

File: scripts/attachSession.sh (hunk 13 above)
Comment (cnishina) [inline scripts/attachSession.sh]: Sure. What folder should I add them to?

**Comment by cnishina** [comment]
See comments above. Thanks!

File: lib/driverProviders/attachSession.js (hunk 14 above)
Comment (sjelin) [inline lib/driverProviders/attachSession.js]: No I mean why not `q(undefined)` instead of `q.fcall(function() {})`?

File: spec/.gitignore [hunk 15]
```diff
@@ -0,0 +1 @@
+attachSessionTemp.js
```
Comment (sjelin) [inline spec/.gitignore]: Is this change still necessary?

File: docs/referenceConf.js [hunk 16]
```diff
@@ -49,6 +49,10 @@ exports.config = {
   // connect to an already running instance of Selenium. This usually looks like
   // seleniumAddress: 'http://localhost:4444/wd/hub'
   seleniumAddress: null,
+  // The selenium session id allows Protractor to attach to an existing selenium
+  // browser session. The selenium session is maintained after the test has
+  // completed. Ignored if seleniumSessionId is null.
```
Comment (juliemr) [inline docs/referenceConf.js]: I think this should be "ignored if seleniumAddress is null"

File: scripts/attachSession.js [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+  };
+  var state = '';
+  var req = http.request(checkOptions, function(res) {
+    res.on('data', function(chunk) {
+      state = JSON.parse(chunk.toString()).state;
+    });
+    res.on('end', function() {
+      if (state === 'success') {
+        var runProtractor = spawn('bin/protractor',
+            ['spec/attachSession.js', '--seleniumSessionId=' + sessionId]);
```
Comment (juliemr) [inline scripts/attachSession.js]: Can we check on the `ChildProcess#close` event and check the exit value of the Protractor process? Then we can also ensure that protractor finishes before stopping the session.

File: scripts/test.js [hunk 18]
```diff
@@ -32,10 +32,11 @@ var passingTests = [
   'node lib/cli.js spec/controlLockConf.js',
   'node lib/cli.js spec/customFramework.js',
   'node lib/cli.js spec/angular2Conf.js',
-  'node scripts/interactive_tests/interactive_test.js',
-  'node scripts/interactive_tests/with_base_url.js',
+  'node scripts/attachSession.js',
+  //'node scripts/interactive_tests/interactive_test.js',
```
Comment (juliemr) [inline scripts/test.js]: Make sure to revert commenting-out in this file before submitting.

File: docs/referenceConf.js [hunk 19]
```diff
@@ -49,6 +49,10 @@ exports.config = {
...
```
Comment (juliemr) [inline docs/referenceConf.js]: Github ate my comment :(

This should be "Ignored if seleniumAddress is null", right?

**Comment by juliemr** [comment]
LGTM after the two cleanup comments

File: docs/referenceConf.js (hunk 19 above)
Comment (cnishina) [inline docs/referenceConf.js]: Yes, changing it.

**Comment by cnishina** [comment]
Closed with 3f3805f.
